File: 101. Symmetric Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
  return root === null || isMirror(root.left, root.right);
};

var isMirror = (left, right) => {
  if (left === null && right === null) return true;
  if (left === null || right === null) return false;
  return (
    left.val === right.val &&
    isMirror(left.left, right.right) &&
    isMirror(left.right, right.left)
  );
};

File: 101. Symmetric Tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}
	return isSame(root.Left, root.Right)
}

func isSame(parent1, parent2 *TreeNode) bool {
	switch {
	case parent1 == nil && parent2 == nil:
		return true
	case parent1 == nil && parent2 != nil, parent1 != nil && parent2 == nil, parent1.Val != parent2.Val:
		return false
	default:
		return isSame(parent1.Left, parent2.Right) && isSame(parent1.Right, parent2.Left)
	}
}